### ISSUE TITLE
Mediapipe ARCamera freeze on scene switch

### ISSUE THREAD

**Issue by patrick508** [homuler/MediaPipeUnityPlugin#373]
Hi,

For our project we use Mediapipe with AR camera input. (I made this pose a while back: https://github.com/homuler/MediaPipeUnityPlugin/issues/343)

So we got it to work, and starting a scene with MediaPipe using this ARCamera works fine. However, if we switch from an ARScene with Mediapipe to another ARScene with mediapipe, the ARCamera from Unity freezes & sends the same image to mediapipe every frame. The ARCamera manager keeps saying Hey i got a new frame! But it's the same one. So no errors are thrown, and mediapipe can detect a pose but it detects a pose on the same image every frame. Together with this the actual camera on screen also freezes. 

I tried switching from an ARScene with mediapipe to a blank AR scene(Only unitys AR camera there), and that works fine. So somewhere in our script it looks like we're still using something that messes up the complete ARCamera for the next scene. 

Does this ring any bell perhaps? Here is the script:

`using Mediapipe.Unity;
using Mediapipe.Unity.PoseTracking;
using System;
using System.Collections;
using ComputerVision.Scripts.Entities;
using ComputerVision.Scripts.Utils;
using ComputerVision.Scripts.Wrappers;
using Entities;
using Unity.Collections;
using Unity.Collections.LowLevel.Unsafe;
using UnityEngine;
using UnityEngine.XR.ARFoundation;
using UnityEngine.XR.ARSubsystems;
using Zenject;

namespace ComputerVision.Scripts.Providers
{
	public class ARPoseProvider: BasePoseProvider, IPoseProvider
    {
        private IPoseTrackingGraphWrapper arCameraPoseTrackingGraph;
        private TextureFramePool arCameraTextureFramePool;
        private WebCamSource arImageSource;
        private PoseTrackingValue currentARCameraPose;
        private PoseData currentPoseData;
        private ARCameraManager arCameraManager;
        private TextureFrame arTextureFrame;

        private Coroutine poseRoutine;
        private Texture2D arTexture;
        private Texture2D arTextureToUpdate;
        
        private bool initialized;

        [Inject]
        public void Inject(ARCameraPoseTrackingGraph arCameraPoseTrackingGraph, WebCamSource arImageSource, ARCameraManager arCameraManager)
        {
            this.arCameraPoseTrackingGraph = arCameraPoseTrackingGraph ?? throw new ArgumentNullException(nameof(arCameraPoseTrackingGraph));
            this.arCameraManager = arCameraManager ?? throw new ArgumentNullException(nameof(arCameraManager));
            this.arImageSource = arImageSource ?? throw new ArgumentNullException(nameof(arImageSource));
        }

        #region Initialization
        public IEnumerator Start()
        {
            AssetLoader.Provide(new StreamingAssetsResourceManager());
            
            StartCoroutine(StartARCameraImagePoseTracking());
            
            yield return StartCoroutine(InitializeInferenceMode());
            
            Debug.Log("PoseProvider is initialized");
            initialized = true;
        }
        #endregion
        
        #region PoseProviding
        private IEnumerator InitializeARCameraSources()
        {
            yield return arImageSource.Play();
        }

        private IEnumerator InitializeARCameraPoseGraphRoutine()
        {
            arCameraPoseTrackingGraph.WaitForInit();
            arCameraPoseTrackingGraph.StartRun(arImageSource);

            yield return null;
        }
        
        private IEnumerator StartARCameraImagePoseTracking()
        {
            yield return StartCoroutine(InitializeARCameraSources());
        
            yield return StartCoroutine(InitializeARCameraPoseGraphRoutine());
            
            arCameraManager.frameReceived += OnARCameraFrameReceived;
        }
        
        private void OnARCameraFrameReceived(ARCameraFrameEventArgs eventArgs)
        {
            arTextureToUpdate = GetCurrentColorTexture();

            if (arTextureToUpdate != null)
            {
                UpdateARPose(arTextureToUpdate);
            }
        }
        
        private void UpdateARPose(Texture2D texture)
        {
            arTextureFrame ??= new TextureFrame(texture.width, texture.height);
            
            arTextureFrame.SetPixels32(texture.GetPixels32());

            arCameraPoseTrackingGraph.AddTextureFrameToInputStream(arTextureFrame).AssertOk();
            currentARCameraPose = arCameraPoseTrackingGraph.FetchNextValue();
            currentPoseData = PoseMapper.GetPoseData(currentARCameraPose);
        }
        
        #endregion
        
        public PoseData GetCurrentPose()
        {
            if (!initialized)
            {
                Debug.LogWarning("PoseProvider is not initialized yet.");
            }
            
            return currentPoseData;
        }

        public PoseTrackingValue GetLowLevelPose()
        {
            if (!initialized)
            {
                Debug.LogWarning("PoseProvider is not initialized yet.");
            }
            
            return currentARCameraPose;
        }

        unsafe Texture2D GetCurrentColorTexture()
        {
            if (!arCameraManager.TryAcquireLatestCpuImage(out XRCpuImage image))
                return null;

            XRCpuImage.ConversionParams conversionParams = new XRCpuImage.ConversionParams
            {
                inputRect = new RectInt(0, 0, image.width, image.height),
                outputDimensions = new Vector2Int(image.width, image.height),
                outputFormat = TextureFormat.RGBA32,
                transformation = XRCpuImage.Transformation.MirrorX
            };

            int size = image.GetConvertedDataSize(conversionParams);
            NativeArray<byte> buffer = new NativeArray<byte>(size, Allocator.Temp);

            image.Convert(conversionParams, new IntPtr(buffer.GetUnsafePtr()), buffer.Length);
            image.Dispose();

            if (arTexture == null)
            {
                arTexture = new Texture2D(conversionParams.outputDimensions.x, conversionParams.outputDimensions.y, conversionParams.outputFormat, false);
            }

            arTexture.LoadRawTextureData(buffer);
            arTexture.Apply();
            buffer.Dispose();

            return arTexture;
        }

        public override void OnDisable()
        {
            arImageSource.Stop();
        }
    }
}`

I know this is our own implementation but perhaps you notice something that we are forgetting to dispose or something that might mess things up.

I tried disposing and stopping all variables but sadly nothing worked.

So to sum up:
Above script works in scene 1, but when we switch to scene 2 which also has this same script, it works for 1 frame, then the screen freezes(unity doesn't and continues to run cause my animations etc continue on) but mediapipe keeps working, it just receives the same texture and thus finds the same pose everytime. The ARCamera manager keeps calling OnFrameReceived() and everything works fine but the texture that comes out of OnFrameReceived is the same everytime....

**Comment by homuler** [comment]
@patrick508 Sorry for the late reply. 

I feel the problem is due to the way `ARCamera` is used, so I'm closing it.
If you think it's not and still need help, please reopen it.

P.S.
At first glance, `BasePoseProvider` and `IPoseProvider` implementations are not provided, so I cannot test the above code.
I think `OnARCameraFrameReceived` should be removed from `arCameraManager.frameReceived` at least, but I don't know the true cause.

**Comment by BilalatNarsun** [comment]
Hey @patrick508 @homuler have you got any resolution with it? I can't figure it our how do you guys feed the AR camera output texture to the media pipe mode. I am trying it with Objectron solution, Can you help me with that?

**Comment by Bruce-yu199** [comment]
@patrick508  Hi, I have the same problem as yours, when i want to switch the scene, it freezes. So i am curious about how you solve it , and how to "undermine the ARCamera's update event"?

**Comment by BilalatNarsun** [comment]
@Bruce-yu199 I got it fixed using an custom script inherited from image source. To feed AR Camera's texture as an input texture.
 
 https://drive.google.com/file/d/1OdW3z8EDnbeeSjmXOp_QFe1fS88N_bK7/view?usp=sharing
 
 For more details contact me on email bilalakram@narsun.pk

**Comment by Bruce-yu199** [comment]
@patrick508 @BilalatNarsun Thanks! I've solved it.

**Comment by Noobno01** [comment]
@BilalatNarsun can you give me access to the drive link?
